dds/DCPS/QOS_XML_Handler/QOS_XML_Loader.cpp

```diff
@@ -8,76 +8,75 @@ OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
 namespace OpenDDS {
 namespace DCPS {
 
-  QOS_XML_Loader::QOS_XML_Loader (void)
+  QOS_XML_Loader::QOS_XML_Loader(void)
   {
   }
 
-  QOS_XML_Loader::~QOS_XML_Loader (void)
+  QOS_XML_Loader::~QOS_XML_Loader(void)
   {
   }
 
 
-  char*
-  QOS_XML_Loader::get_xml_file_name (const char * qos_profile)
+  ACE_TCHAR*
+  QOS_XML_Loader::get_xml_file_name(const ACE_TCHAR * qos_profile)
   {
     if (qos_profile)
       {
-        char* buf = ACE_OS::strdup (qos_profile);
-        ACE_Tokenizer_T<char> tok (buf);
-        tok.delimiter_replace ('#', 0);
-        const char * file_name = tok.next ();
+        ACE_TCHAR* buf = ACE_OS::strdup(qos_profile);
+        ACE_Tokenizer_T<ACE_TCHAR> tok(buf);
+        tok.delimiter_replace('#', 0);
+        const ACE_TCHAR * file_name = tok.next();
 
         if (file_name == 0)
           {
-            ACE_OS::free (buf);
+            ACE_OS::free(buf);
             if (DCPS_debug_level > 5)
               {
-                ACE_ERROR ((LM_ERROR,
-                              "get_xml_file_name <%C> - "
-                              "Error: malformed qos_profile. Expected format: "
-                              "<xml_file_base_name>#<profile_name>\n",
-                              qos_profile));
+                ACE_ERROR((LM_ERROR,
+                           "get_xml_file_name <%s> - "
+                           "Error: malformed qos_profile. Expected format: "
+                           "<xml_file_base_name>#<profile_name>\n",
+                           qos_profile));
               }
             return 0;
           }
 
-        char * ret = (char*)ACE_OS::malloc (ACE_OS::strlen (file_name) + 5);
-        ret = ACE_OS::strcpy (ret, file_name);
-        ret = ACE_OS::strcat (ret, ".xml");
-        ACE_OS::free (buf);
-        return ret;
+        ACE_TString ret(file_name);
+        ret += ACE_TEXT(".xml");
+        ACE_OS::free(buf);
+        return ret.rep();
```

Comment:
Looks like it's returning a pointer that will be freed before it can be used.

Reply:
http://www.dre.vanderbilt.edu/Doxygen/5.7.9/html/ace/a00651.html#a2bca515bf58fa5e32c67561f3a141a83

"This method allocates memory for a copy of the string and returns a pointer to the new area. The caller is responsible for freeing the memory when finished"

Reply:
OK, I thought that may have been the `c_str` equivalent.  Looking at this one had me thinking we should update the interface to return objects instead of pointers, but I'm not sure if that breaks anything for DDS4CCM.

Reply:
Yes, that very likely will need some dds4ccm changes, feel free to update dds4ccm when required

Reply:
Where are we on this? It's doing what it's always done, but now it compiles with wchar support. Am I doing more here?

Reply:
ACE's doxygen:
```
   * This method allocates memory for a copy of the string and returns
   * a pointer to the new area. The caller is responsible for freeing
   * the memory when finished; use delete []
```
so the caller of this is now tasked with remembering to use `delete[]` instead of some other function/operator.  I don't think that's what we want.

Now that I'm in front of a full development environment looking at how this is used, I think we should change these two private methods.  This is what `std::basic_string` (with its various template arguments) is for.